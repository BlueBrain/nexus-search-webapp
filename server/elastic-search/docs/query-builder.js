/**
 *
 *
 * @param {string} query
 * @returns {object} an elastic search query object
 */
function makeDocsQuery(
  query = { filter: null, type: null, q: null, sort: null }
) {
  let params = {
    query: {
      bool: {
        must: []
      }
    }
  };
  let { sort, filter, type, q } = query;
  // TODO mop this up by using MAPPING or SCHEMA to assemble sorting
  if (sort) {
    sort = JSON.parse(sort);
    if (sort.field) {
      if (sort.field === "@type") {
        let field = sort.field + ".raw";
        params.sort = [
          {
            [field]: {
              order: sort.order
            }
          }
        ];
      } else {
        let field = sort.field + ".label.raw";
        params.sort = [
          {
            [field]: {
              order: sort.order,
              nested: {
                path: sort.field
              }
            }
          }
        ];
      }
    }
  }
  if (q) {
    params.query.bool.must.push({
      query_string: {
        query: `(${query.q}* OR ${query.q}~)`
      }
    });
  }
  if (type) {
    params.query.bool.must.push({
      term: { "@type.raw": query.type }
    });
  }
  if (filter) {
    filter = JSON.parse(filter);
    // MUST is an AND, we do it for terms across filter sets
    let must = Object.keys(filter).map(key => {
      // SHOULD is an OR, we do it for terms in the same filter set
      let should = filter[key].map(filterTerm => {
        let propertyName = `${key}.raw`;
        return { term: { [propertyName]: filterTerm } };
      });

      // For nested queries, every part of the path needs to contain
      // its ancestors, as in "grandparent.parent.value"
      let path = key.split(".").reduce((previous, current) => {
        const parentLabel = previous[previous.length - 1] || null;
        const currentLabel = parentLabel
          ? `${parentLabel}.${current}`
          : current;
        previous.push(currentLabel);
        return previous;
      }, []);

      return path.reverse().reduce((memo, level, index) => {
        if (index === 0) {
          memo = {
            bool: {
              should
            }
          };
        } else {
          memo = {
            nested: {
              path: level,
              query: memo
            }
          };
        }
        return memo;
      }, {});
    });
    if (must.length) {
      params.query.bool.must.push({
        bool: {
          must
        }
      });
    }
  }
  console.log(JSON.stringify(params, null, 2))
  return params;
}

export default makeDocsQuery;
