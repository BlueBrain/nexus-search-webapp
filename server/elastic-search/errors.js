class ElasticSearchError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ElasticSearchError';
  }
}

export {
  ElasticSearchError
}