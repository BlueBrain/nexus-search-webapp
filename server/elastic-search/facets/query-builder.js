import mappingToAggs from "./mappings-to-aggs";

async function getAggsFromMapping (client, index) {
  let properties = await client.indices.getMapping({
    index,
    type: "doc"
  })
  properties = properties[index].mappings.doc.properties;
  return mappingToAggs(properties)
}

/**
 *
 *
 * @param {object} nexusSearchQueryObject
 * @returns {object} an elastic search query object
 */
async function makeFacetQuery({ q, type }, client, index) {
  let params = {
    aggs: await getAggsFromMapping(client, index)
  };
  if (type) {
    params.query = {
      match: {
        "@type.raw": type
      }
    };
  }
  return params;
}

export default makeFacetQuery;
