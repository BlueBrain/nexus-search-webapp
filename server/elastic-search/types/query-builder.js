const typeAggs = {
  "@types": {
    terms: {
      field: "@type.raw"
    }
  }
};

/**
 * make a query_string es query object with a given string input
 *
 * @param {string} textQuery
 * @returns {object} an elastic search query_string object
 */
function makeQueryStringWithText (textQuery) {
  return {
    query_string: {
        query: `(${textQuery}* OR ${textQuery}~)`
    }
  };
}

/**
 *
 *
 * @param {string} textQuery
 * @returns {object} an elastic search query object
 */
function makeTypeQuery(textQuery) {
  let params = {
    aggs: typeAggs
  };
  if (textQuery) {
    params.query = makeQueryStringWithText(textQuery)
  }
  return params;
}

export default makeTypeQuery;
