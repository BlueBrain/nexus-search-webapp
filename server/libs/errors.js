// For when ES breaks
class ElasticSearchError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ElasticSearchError';
  }
}

// For when Resource API breaks
class ResourceError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ResourceError';
  }
}

// For when QB breaks
class QueryBuilderError extends Error {
  constructor(message) {
    super(message);
    this.name = 'QueryBuilderError';
  }
}

// For when the server is not configured correctly
class InvalidConfigError extends Error {
  constructor(message) {
    super(message);
    this.name = 'InvalidConfigError';
  }
}

// For when the token is invalid
class UnauthorizedError extends Error {
  constructor(message) {
    super(message);
    this.name = 'UnauthorizedError';
  }
}

// For when a request to an external service is broken
class ExternalServiceError extends Error {
  constructor(message) {
    super(message);
    this.name = 'ExternalServiceError';
  }
}

// For when a request to an external service returns 404
class NotFoundError extends Error {
  constructor(message) {
    super(message);
    this.name = 'NotFound';
  }
}

export {
  ElasticSearchError,
  QueryBuilderError,
  UnauthorizedError,
  InvalidConfigError,
  ExternalServiceError,
  NotFoundError
}