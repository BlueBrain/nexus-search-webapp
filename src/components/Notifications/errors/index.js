import React from "react";
import SVG from "react-svg";
import icons from "./../../Icons";


const description = <div>
  <p style={{fontSize: "1em"}}>We're sorry, but this seems to be an error! If this isn't the first time you have seen this, consider letting us know.</p>
  <a target="_blank" href="https://github.com/BlueBrain/nexus-search-webapp/issues/new">post an issue to github</a>
</div>

const connectionError = {
  message: "Something went when reaching out to the backend.",
  description,
  icon: <div style={{ width: "2em", marginRight: "1em"}}>
  <SVG
    path={icons.bug}
    svgClassName="notification-svg error"
  />
</div>,
}

const somethingWentWrongError = {
  message: "Oh no! Something went wrong.",
  description,
  icon: <div style={{ width: "2em", marginRight: "1em"}}>
  <SVG
    path={icons.bug}
    svgClassName="notification-svg error"
  />
</div>,
}

export default {
  connectionError,
  somethingWentWrongError
}