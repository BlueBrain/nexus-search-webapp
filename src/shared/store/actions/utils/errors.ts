import { HTTP_STATUS_TYPE_KEYS, HTTP_STATUSES } from './statusCodes';

export class RequestError extends Error {
  readonly message: string;
  readonly type: string;
  readonly code: number;
  constructor(
    message: string,
    type: string = HTTP_STATUS_TYPE_KEYS.BAD_REQUEST
  ) {
    super(message);
    this.message = message;
    this.type = type;
    this.code = HTTP_STATUSES[type].code;
  }
}

export const formatError = (error: Error) => {
  const errorType = error.message.replace(/ /g, '_').toUpperCase();
  const httpErrorType = HTTP_STATUSES[errorType];
  if (httpErrorType) {
    return new RequestError(error.message, httpErrorType.type);
  }
  return new RequestError(error.message);
};
