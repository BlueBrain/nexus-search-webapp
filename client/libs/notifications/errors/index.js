import { notification } from 'antd';
import Notifications from "../../../components/Notifications";

const DEFAULT_ERROR_NOTIFICATION_CONFIG = {
  duration: 0,
}

export default defaultConfig => {
  const errorConfig = Object.assign(defaultConfig, DEFAULT_ERROR_NOTIFICATION_CONFIG);
  return {
    connectionError () {
      const { message, description, icon } = Notifications.errors.connectionError;
      notification.open({
        message,
        description,
        icon,
        ...errorConfig
      });
    },
    somethingWentWrongError () {
      const { message, description, icon } = Notifications.errors.somethingWentWrongError;
      notification.open({
        message,
        description,
        icon,
        ...errorConfig
      });
    }
  }
}