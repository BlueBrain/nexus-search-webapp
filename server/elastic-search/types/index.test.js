import typesFactory from "./index";
// * necessary to mock custom modules
import * as queryBuilder from "./query-builder"
import * as Errors from "../errors";
const fakeNormalizer = docs => docs

describe("typesFactory()", () => {
  it("returns a function", () => {
    expect(typeof typesFactory()).toBe("function");
  });

  describe("fetchTypes()", () => {
    it("returns a promise", async () => {
      const fakeTypesResults = {
        docs: {
          aggrigations: {
            "@types": {
              buckets: []
            }
          }
        }
      }
      expect.assertions(1);
      const fakeClient = {
        search: () => Promise.resolve(fakeTypesResults)
      };
      const fetchTypes = typesFactory(fakeClient, "my_index", fakeNormalizer);
      const data = await fetchTypes();
      return expect(data).toBe(fakeTypesResults);
    });

    it("calls the es client search function", async () => {
      const fakeClient = {
        search: jest.fn()
      };
      const fetchTypes = typesFactory(fakeClient, "my_index", fakeNormalizer);
      fetchTypes();
      await expect(fakeClient.search.mock.calls.length).toBe(1);
    });

    it("it calls es client search function with params index, type, and doc", () => {
      const fakeClient = {
        search: jest.fn()
      };
      const fetchTypes = typesFactory(fakeClient, "my_index", fakeNormalizer);
      fetchTypes();
      expect(fakeClient.search.mock.calls[0][0]).toMatchSnapshot()
    });

    it("it calls the types queryBuilder", () => {
      queryBuilder.default = jest.fn();
      const fakeClient = {
        search: jest.fn()
      };
      const fetchTypes = typesFactory(fakeClient, "my_index", fakeNormalizer);
      fetchTypes();
      expect(queryBuilder.default).toBeCalled();
    });

    it("makes an ElasticSearchError if serach fails", async () => {
      const fakeClient = {
        search: () => Promise.reject("I exploded")
      };
      const fetchTypes = typesFactory(fakeClient, "my_index", fakeNormalizer);
      expect.assertions(2);
      await expect(fetchTypes()).rejects.toBeInstanceOf(Error);
      await expect(fetchTypes()).rejects.toHaveProperty("name", "ElasticSearchError");
    })

    it("makes an QueryBuilderError if serach fails", async () => {

    })
  });
});
