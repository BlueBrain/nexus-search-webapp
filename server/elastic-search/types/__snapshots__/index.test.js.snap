// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`typesFactory() fetchTypes() it calls es client search function with params index, type, and doc 1`] = `
Object {
  "body": Object {
    "aggs": Object {
      "@types": Object {
        "terms": Object {
          "field": "@type.raw",
        },
      },
    },
  },
  "index": "my_index",
  "type": "doc",
}
`;
